src/syntax_analysis.rs

```diff
@@ -0,0 +1,192 @@
+
+        dependency_to_syntax.insert(Dependency::ByName(name.clone()), syntax);
+        dependency_to_syntax.insert(Dependency::ByScope(*scope), syntax);
+    }
+
+    (syntax_to_dependencies, dependency_to_syntax)
+}
+
+/// Gets what external dependencies a given [SyntaxDefinition] has.
+/// An external dependency is another `.sublime-syntax` file.
+/// It does that by looking for variants of the following YAML patterns:
+/// - `include: C.sublime-syntax`
+/// - `embed: scope:source.c`
+fn dependencies_for_syntax(syntax: &SyntaxDefinition) -> Vec<Dependency> {
+    let mut dependencies: Vec<Dependency> = syntax
+        .contexts
+        .values()
+        .map(|context| &context.patterns)
+        .flatten()
+        .map(dependencies_from_pattern)
+        .flatten()
+        .collect();
+
+    // No need to track a dependency more than once
+    dependencies.dedup();
+
+    dependencies
+}
+
+fn dependencies_from_pattern(pattern: &Pattern) -> Vec<Dependency> {
+    match *pattern {
+        Pattern::Match(MatchPattern {
+            operation: MatchOperation::Push(ref context_references),
+            ..
+        }) => context_references
+            .iter()
+            .map(dependency_from_context_reference)
+            .collect(),
+        Pattern::Include(ref cr) => vec![dependency_from_context_reference(cr)],
+        _ => vec![],
+    }
+    .into_iter()
+    .flatten()
+    .collect()
+}
+
+fn dependency_from_context_reference(context_reference: &ContextReference) -> Option<Dependency> {
+    match &context_reference {
+        ContextReference::File { ref name, .. } => Some(Dependency::ByName(name.clone())),
+        ContextReference::ByScope { ref scope, .. } => Some(Dependency::ByScope(*scope)),
+        _ => None,
+    }
+}
+
+/// Helper to construct a [SyntaxSetBuilder] that contains only [SyntaxDefinition]s
+/// that have dependencies among them.
+struct SyntaxSetDependencyBuilder {
+    syntax_set_builder: SyntaxSetBuilder,
+}
+
+impl SyntaxSetDependencyBuilder {
+    fn new() -> Self {
+        SyntaxSetDependencyBuilder {
+            syntax_set_builder: SyntaxSetBuilder::new(),
+        }
+    }
+
+    /// Add a [SyntaxDefinition] to the underlying [SyntaxSetBuilder].
+    /// Also resolve any dependencies it has and add those [SyntaxDefinition]s too.
+    /// This is a recursive process.
+    fn add_with_dependencies(
+        &mut self,
+        syntax: &SyntaxDefinition,
+        syntax_to_dependencies: &SyntaxToDependencies,
+        dependency_to_syntax: &DependencyToSyntax,
+    ) {
+        let name = &syntax.name;
+        if self.is_syntax_already_added(name) {
+            return;
+        }
+
+        let dependencies = syntax_to_dependencies.get(name);
+        if dependencies.is_none() {
+            if VERBOSE {
+                eprintln!("WARNING: Unknown dependencies for {}", name);
+            }
+            return;
+        }
+
+        self.syntax_set_builder.add(syntax.clone());
```

Comment:
we should add the source syntax *before* we return early, so that there is always one SyntaxDefinition to build.

It is an edge case, because it shall never happen, but doesn't hurt to be robust.